File: com.ripleydynamics.filament-dialin/lib/tower.lua
-- Stacked-section tower builder shared by the temperature, flow, volumetric
-- and stringing commands.
--
-- A tower is one solid block (width x depth x total height) with, per section:
--   * an optional custom G-code entry emitted on the section's first layer,
--   * an optional modifier volume carrying per-region settings,
--   * a label engraved on the front (-Y) face.
-- The printer tag is engraved on the back face of the plinth.

local M = {}

M.MODIFIER_MARGIN = 2 -- mm the modifier boxes overhang the body on every side

-- Z at which a custom G-code entry lands on the first layer of a section that
-- starts at z0. PrusaSlicer assigns an entry to the first layer whose print_z
-- is at or above the entry, so half a layer up is unambiguous whatever the
-- first-layer height is.
function M.gcode_z(z0, layer_height)
    return z0 + layer_height * 0.5
end

function M.label_line_height(section_h)
    return math.max(2.5, math.min(7, section_h * 0.55))
end

-- spec:
--   width, depth        footprint (X, Y) in mm
--   base_height         plinth under the first section (0 allowed)
--   section_height      requested section height, rounded to whole layers
--   layer_height        current layer height
--   sections            list of { label = string?, gcode = string?, params = table? }
--   tag                 text for the back of the plinth (optional)
--   extra_x             extra +X reach the modifiers must cover (wings)
-- Returns an add_object definition with extra fields:
--   total_height, section_height, base_height, section_z (list of z0)
function M.build(bed, spec)
    local util = require("lib/util")
    local label = require("lib/label")

    local n = #spec.sections
    assert(n >= 1, "a tower needs at least one section")
    local lh = spec.layer_height
    assert(lh and lh > 0, "layer height must be positive")

    local w, d = spec.width, spec.depth
    local section_h = util.align(spec.section_height, lh, 2)
    local base_h = spec.base_height > 0 and util.align(spec.base_height, lh, 1) or 0
    local total = base_h + n * section_h
    local margin = M.MODIFIER_MARGIN
    local extra_x = spec.extra_x or 0

    local volumes = {}
    local section_z = {}

    -- Custom G-code is a per-bed list; replace it wholesale so stale entries
    -- from a previous tower cannot linger.
    api.project:clear_layer_custom_steps(bed)

    for i, s in ipairs(spec.sections) do
        local z0 = base_h + (i - 1) * section_h
        section_z[i] = z0

        if s.gcode and s.gcode ~= "" then
            api.project:insert_layer_custom_gcode(bed, M.gcode_z(z0, lh), s.gcode)
        end

        if s.params then
            volumes[#volumes + 1] = {
                mesh = api.make_cube(w + 2 * margin + extra_x, d + 2 * margin, section_h),
                type = VolumeType.Modifier,
                translate = { x = -margin, y = -margin, z = z0 },
                params = util.merge(s.params),
            }
        end

        if s.label and s.label ~= "" then
            volumes[#volumes + 1] = label.front {
                text = s.label,
                x = w / 2,
                z = z0 + section_h / 2,
                face_y = 0,
                line_height = M.label_line_height(section_h),
                max_width = w - 3,
                max_height = section_h - 1.5,
            }
        end
    end

    if spec.tag and spec.tag ~= "" and base_h >= 3 then
        volumes[#volumes + 1] = label.back {
            text = spec.tag,
            x = w / 2,
            z = base_h / 2,
            face_y = d,
            line_height = math.min(6, base_h * 0.6),
            max_width = w - 3,
            max_height = base_h - 1,
        }
    end

    return {
        mesh = api.make_cube(w, d, total),
        other_volumes = volumes,
        total_height = total,
        section_height = section_h,
        base_height = base_h,
        section_z = section_z,
    }
end

-- A slab attached to the +X face of the body, tilted so its underside is an
-- overhang of `angle_deg` from horizontal. The slab is a cube rotated about
-- its own corner (the volume origin), so the geometry is fully predictable:
-- bottom edge rises from (x_face, z0) at the requested angle.
-- Returns the volume definition and how far (mm) the wing reaches in +X.
function M.wing(o)
    local angle = o.angle_deg
    local t = o.thickness or 2
    local rad = math.rad(angle)
    local rise_available = o.section_height - 0.5 - t * math.cos(rad)
    assert(rise_available > 1, "section too short for an overhang wing")
    local length = rise_available / math.sin(rad)
    local wing_depth = o.wing_depth or o.depth * 0.6
    local def = {
        mesh = api.make_cube(length, wing_depth, t),
        type = VolumeType.Solid,
        rotate = { y = -angle },
        translate = { x = o.x_face, y = (o.depth - wing_depth) / 2, z = o.z0 },
    }
    return def, length * math.cos(rad)
end

return M

File: port29/tower.py
"""Stacked-section tower builder, port of lib/tower.lua."""

import math

from . import geometry as g
from .common import align
from .threemf import MODEL_PART, MODIFIER, NEGATIVE_VOLUME, Object3mf, Volume

MODIFIER_MARGIN = 2


def gcode_z(z0, layer_height):
    """Half a layer above the section start: lands on its first layer."""
    return z0 + layer_height * 0.5


def label_line_height(section_h):
    return max(2.5, min(7.0, section_h * 0.55))


def build(project, spec):
    """spec keys: width, depth, base_height, section_height, layer_height,
    sections (list of dicts with optional label, gcode, params), tag, extra_x,
    name. Adds the custom G-code to `project`; returns (Object3mf, dims) where
    dims has total_height, section_height, base_height, section_z."""
    sections = spec["sections"]
    if not sections:
        raise ValueError("a tower needs at least one section")
    lh = spec["layer_height"]
    w, d = float(spec["width"]), float(spec["depth"])
    section_h = align(spec["section_height"], lh, 2)
    base_h = align(spec["base_height"], lh, 1) if spec.get("base_height", 0) > 0 else 0.0
    total = base_h + len(sections) * section_h
    margin, extra_x = MODIFIER_MARGIN, spec.get("extra_x", 0)

    volumes = [Volume(g.box(w, d, total), MODEL_PART, spec.get("name", "tower"))]
    section_z = []
    for i, s in enumerate(sections):
        z0 = base_h + i * section_h
        section_z.append(z0)
        if s.get("gcode"):
            project.add_custom_gcode(gcode_z(z0, lh), s["gcode"])
        if s.get("params"):
            volumes.append(Volume(g.box(w + 2 * margin + extra_x, d + 2 * margin, section_h).translate(-margin, -margin, z0),
                                  MODIFIER, f"band {i + 1}", dict(s["params"])))
        if s.get("label"):
            volumes.append(Volume(g.label_front(s["label"], w / 2, z0 + section_h / 2, 0.0, label_line_height(section_h), w - 3, section_h - 1.5),
                                  NEGATIVE_VOLUME, f"label {s['label']}"))
    if spec.get("tag") and base_h >= 3:
        volumes.append(Volume(g.label_back(spec["tag"], w / 2, base_h / 2, d, min(6.0, base_h * 0.6), w - 3, base_h - 1),
                              NEGATIVE_VOLUME, "tag"))
    obj = Object3mf(spec.get("name", "tower"), volumes)
    return obj, {"total_height": total, "section_height": section_h, "base_height": base_h, "section_z": section_z}


def wing(x_face, depth, z0, section_height, angle_deg, thickness=2.0, wing_depth=None):
    """Overhang slab on the +X face, underside at angle_deg from horizontal.
    Returns (Volume, reach in +X)."""
    rad = math.radians(angle_deg)
    rise_available = section_height - 0.5 - thickness * math.cos(rad)
    if not rise_available > 1:
        raise ValueError("section too short for an overhang wing")
    length = rise_available / math.sin(rad)
    wd = wing_depth if wing_depth is not None else depth * 0.6
    mesh = g.box(length, wd, thickness).rotate_y(-angle_deg).translate(x_face, (depth - wd) / 2, z0)
    return Volume(mesh, MODEL_PART, f"wing {angle_deg:g}"), length * math.cos(rad)
